liquidity/loopout_builder.go

```diff
@@ -0,0 +1,160 @@
+package liquidity
+
+import (
+	"context"
+
+	"github.com/btcsuite/btcutil"
+	"github.com/lightninglabs/loop"
+	"github.com/lightninglabs/loop/labels"
+	"github.com/lightninglabs/loop/loopdb"
+	"github.com/lightninglabs/loop/swap"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+// Compile-time assertion that loopOutBuilder satisfies the swapBuilder
+// interface.
+var _ swapBuilder = (*loopOutBuilder)(nil)
+
+func newLoopOutBuilder(cfg *Config) *loopOutBuilder {
+	return &loopOutBuilder{
+		cfg: cfg,
+	}
+}
+
+type loopOutBuilder struct {
+	// cfg contains all the external functionality we require to create
+	// swaps.
+	cfg *Config
+}
+
+// swapType returns the swap type that the builder is responsible for creating.
+func (b *loopOutBuilder) swapType() swap.Type {
+	return swap.TypeOut
+}
+
+// maySwap checks whether we can currently execute a swap, examining the
+// current on-chain fee conditions against relevant to our swap type against
+// our fee restrictions.
+//
+// For loop out, we check whether the fees required for our on-chain sweep
+// transaction exceed our fee limits.
+func (b *loopOutBuilder) maySwap(ctx context.Context, params Parameters) error {
+	estimate, err := b.cfg.Lnd.WalletKit.EstimateFee(
+		ctx, params.SweepConfTarget,
+	)
+	if err != nil {
+		return err
+	}
+
+	return params.FeeLimit.mayLoopOut(estimate)
+}
+
+// inUse examines our current swap traffic to determine whether we should
+// suggest the builder's type of swap for the peer and channels suggested.
+func (b *loopOutBuilder) inUse(traffic *swapTraffic, peer route.Vertex,
+	channels []lnwire.ShortChannelID) error {
+
+	for _, chanID := range channels {
+		lastFail, recentFail := traffic.failedLoopOut[chanID]
+		if recentFail {
+			log.Debugf("Channel: %v not eligible for suggestions, "+
+				"was part of a failed swap at: %v", chanID,
+				lastFail)
+
+			return newReasonError(ReasonFailureBackoff)
+		}
+
+		if traffic.ongoingLoopOut[chanID] {
+			log.Debugf("Channel: %v not eligible for suggestions, "+
+				"ongoing loop out utilizing channel", chanID)
+
+			return newReasonError(ReasonLoopOut)
+		}
+	}
+
+	if traffic.ongoingLoopIn[peer] {
+		log.Debugf("Peer: %x not eligible for suggestions ongoing "+
+			"loop in utilizing peer", peer)
+
+		return newReasonError(ReasonLoopIn)
+	}
+
+	return nil
+}
+
+// buildSwap creates a swap for the target peer/channels provided. The autoloop
+// boolean indicates whether this swap will actually be executed, because there
+// are some calls we can leave out if this swap is just for a dry run.
```

Comment:
Dryrun would flip the meaning of the bool, and since we have it in other parts of the `liquidity` package as `autoloop` I'm going to leave it? Reasoning for `autoloop` is that it's opt-in, so made sense to have a bool that defaulted to false.

Reply:
Agree, though I think the returned suggestion is kind of a dry run depending on use. We could avoid generating a new wallet though. Maybe some more clarification in the godoc can help understanding what dry-run means in this context.